Migrate from Schema.org to JSON-LD

juliannatetreault opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

We currently use schema.org to structure DEV’s data. This PR removes schema.org and its `itemprop`s from the templates that used it and replaces it with the JSON-LD data structure.

To test these changes, I have been using Google's Structured Data Testing Tool, which can be found here: https://search.google.com/structured-data/testing-tool/u/0/, in conjunction with the specs that I've written around these changes.

## Related Tickets & Documents

Closes #6798 

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![SNL Google](https://media.giphy.com/media/vxRJWOAghqjXG/giphy.gif)

benhalpern commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,25 +7,35 @@
+    "@context": "http://schema.org",
+    "@type": "Person",
+    "mainEntityOfPage": "True",
+    "url": "https://dev.to/<%= @user.username %>",
```
This should be `user_url` similar to `article_url` used elsewhere in the PR.

juliannatetreault replied:
Ah, okay, I'll change it to, `<%= user_url(@user) %>`, right now! Good catch. 👀

benhalpern commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,41 @@
+      "@context": "http://schema.org",
+      "@type": "Organization"
+    },
+    "logo": {
```
Looks like logo should maybe be `within` publisher here?

https://developers.google.com/search/docs/data-types/article

Like this?

```
       "publisher": {
        "@type": "Organization",
        "name": "Google",
        "logo": {
          "@type": "ImageObject",
          "url": "https://google.com/logo.jpg"
        }
      },
```

juliannatetreault replied:
Great catch! I've made this change.

benhalpern commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,41 @@
+    },
+    "datePublished": "<%= @article.published_timestamp %>",
+    "dateModified": ""<%= @article.edited_at&.utc&.iso8601 %>"",
+    "articleBody": "<%= @article.processed_html.html_safe %>"
```
I'm seeing conflicting info about whether `articleBody` should be included.

https://support.google.com/webmasters/thread/8926582?hl=en

juliannatetreault replied:
Good catch! After looking into it some more, I have removed `articleBody`. 😄

benhalpern commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,29 @@
       </div>
       <div class="profile-details">
         <h1>
-          <span itemprop="name"><%= @user.name %><%= render "shared/pro_checkmark", style: nil, width: nil %></span>
```
Looks like this deletion deleted a little too much. The `span` is no longer needed, but the content within should stay (You can see on the profile page, the name is no longer there as it should be.

benhalpern commented on app/views/users/show.html.erb:
```diff
@@ -44,9 +58,7 @@
             </span>
           <% end %>
         </h1>
-        <p class="profile-description" itemprop="description">
```
Same here. The content should still be there, we just need to get rid of the `itemprop="description"` part.

benhalpern commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,25 +7,35 @@
     </div>
     <div class="profile-details">
       <h1 style="color:<%= HexComparer.new([user_colors(@user)[:bg], user_colors(@user)[:text]]).brightness(0.78) %>">
-        <span itemprop="name"><%= sanitize @user.name %></span>
```
Same as otherwise described. Content itself should still be present.

benhalpern commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,25 +7,35 @@
           <button id="user-follow-butt" class="cta follow-action-button user-profile-follow-button" style="color:<%= user_colors(@user)[:text] %>;background-color:<%= user_colors(@user)[:bg] %>" data-info='{"id":<%= @user.id %>,"className":"<%= @user.class.name %>"}'>&nbsp;</button>
         </span>
       </h1>
-      <p class="profile-description" itemprop="description">
```
Same as otherwise described. Content itself should still be present.

benhalpern commented on app/views/articles/show.html.erb:
```diff
@@ -147,7 +166,7 @@
             <em> and <b><a href="<%= @third_user.path %>"><%= @third_user.name %></a></b></em>
           <% end %>
           <% if should_show_updated_on?(@article) %>
-            <span><em>Updated on <time itemprop="dateModified" datetime="<%= @article.edited_at&.utc&.iso8601 %>"><%= @article.edited_at&.strftime("%b %d, %Y") %></time></em></span>
```
`itemprop="dateModified"` should be deleted, but `datetime="<%= @article.edited_at&.utc&.iso8601 %>"` should stay.

benhalpern commented:
Check all the spots where content was deleted and make sure you're only removing the `itemprop` etc. parts, and not the rest. It's only those attributes that need to disappear.

benhalpern commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,25 +7,35 @@
+    "mainEntityOfPage": "True",
+    "url": "<%= user_url(@user) %>",
+    "image": {
+      "@type": "ImageObject"
```
Based on this example, it looks like the `image` attribute of a `Person` type should just be a URL...

https://jsonld.com/person/

benhalpern commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,25 +7,35 @@
+<script type="application/ld+json">
+  {
+    "@context": "http://schema.org",
+    "@type": "Person",
```
Hmmmm.... Looks like `_profile_header.html.erb` is only used in "organization" profiles... which is sort of odd. But this should probably be changed to data for organizations.

juliannatetreault replied:
Hmmm. I've made the change from "Person" to "Organization" here, but it feels weird so let me know how it looks once I push the changes up!

benhalpern commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,40 @@
+        "height": "192"
+      }
+    },
+    "name": "<%= ApplicationConfig["COMMUNITY_NAME"] %> Community",
```
This should probably go _within_ the "publisher" info object, eh?

benhalpern commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,29 @@
+
+<script type="application/ld+json">
+  {
+    "@context": "http://schema.org",
```
I think we have a bunch of other possibly valuable info that could fit a person object similar to what we have here...

https://jsonld.com/person/

For example, if someone has a public email, we could have that, if they have a job title listed we could use those.

Looks like we could also get `sameAs` info by listing social profile links.

juliannatetreault replied:
Okay, I've made some adjustments to include more information for the Person object. I attempted to add further information including employer name and url as well as a User's education, but haven't been able to get it to work/render without an issue yet--I will continue trying to build this information out. 😊

benhalpern replied:
Great. I think what we have looks good and we can make some of those additional fields a follow-on sprint.

One thing though, I think we should probably not include the key at all if it's not present.

This will look kind of messy, but I think something like...

```
  <% if @user.employment_title.present? %>
     "jobTitle": "<%= @user.employment_title %>",
  <% end %>
```
on any column that is optional.

Let's also not leak the user's email if they haven't made it public. We have a column for that, so so like...

```
  <% if @user.email_public %>
     "email": "<%= @user.email %>",
  <% end %>
```

benhalpern commented:
This is really coming together nicely. Looks like there's a conflict to resolve.

Good job making turnarounds on the feedback. I'll give it another pass tomorrow.

In the meantime I think some basic `request` specs which test for simple `expect(response.body).to include "SOME BASIC JSON+LD EXPECTED STRING"` would be good for this PR to ensure it basically acts as expected.

(E.g. in the `stories_show_spec` etc.)

rhymes commented on app/views/articles/_video_player.html.erb:
```diff
@@ -1,18 +1,26 @@
   <div id="video-player-<%= article.id %>" class="video-player"></div>
 </div>
 
+<script type="application/ld+json">
```
do you know if this should technically be in the document's `<head>` ?

juliannatetreault replied:
I do know! 😄 It is okay to place the JSON-LD in both the document's `<head>` and `<body>`, but the `<head>` is recommended.

![Screen Shot 2020-04-03 at 1 29 47 PM](https://user-images.githubusercontent.com/32834804/78397882-470d4500-75af-11ea-9600-3814b83787a2.png)

rhymes commented on app/views/articles/_video_player.html.erb:
```diff
@@ -1,18 +1,26 @@
+    "@context": "https://schema.org",
+    "@type": "VideoObject",
+    "url": "<%= article.video_source_url %>",
+    "uploadDate": "<%= article.published_at %>",
```
Maybe we can change this to `article.published_at&.rfc3339` to render a RFC compliant date string, what do you think?

BTW is the name arbitrary? If so maybe we can change it to `datePublished` like we do in `article/show.html.rb`, don't know 🤷‍♂

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,36 @@
+      "logo": {
+        "@context": "http://schema.org",
+        "@type": "ImageObject",
+        "url": "<%= cloudinary(SiteConfig.logo_png, "png") %>",
+        "width": "192",
+        "height": "192"
```
by default the logo is 512x512 so we should probably resize it to 192 with `cloudinary(SiteConfig.logo_png, 192)`. The second argument is the width

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,36 @@
+      }
+    },
+    "headline": "<%= @article.title %>",
+    "author": {
```
should we add the organization info it is there? not a blocker though :D

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -120,25 +140,20 @@
             <a href="http://github.com/<%= @user.github_username %>"><%= image_tag_or_inline_svg_tag "github", width: 18, height: 18 %></a>
           <% end %>
           <% if @article.published_timestamp.present? %>
-            <time itemprop="datePublished" datetime="<%= @article.published_timestamp %>"><%= @article.readable_publish_date %></time>
```
datetime here is important because it's processed by the client side :)

https://github.com/thepracticaldev/dev.to/blob/c1cf1df7a93afa78e47a45701f3dd818dd6d8d7a/app/assets/javascripts/initializers/initializeArticleDate.js#L6-L12

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -120,25 +140,20 @@
-            <meta itemprop="url" content="<%= user_url(@user) %>">
-            <a href="/<%= @user.username %>" class="author">
-              <img class="profile-pic" src="<%= ProfileImage.new(@user).get(width: 50) %>" alt="<%= @user.username %> profile image" />
-              <span itemprop="name"><%= @user.name %></span>
```
the span itself should stay in, otherwise we don't show the username anymore in the article page

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -85,33 +115,23 @@
-      <meta itemprop="image" content="<%= article_social_image_url(@article) %>">
-      <div itemprop="publisher" itemscope itemtype="https://schema.org/Organization">
-        <div itemprop="logo" itemscope itemtype="https://schema.org/ImageObject">
-          <meta itemprop="url" content="<%= cloudinary(SiteConfig.logo_png, 192, "png") %>">
```
see my other note about resizing the cloudinary URL

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -120,25 +140,20 @@
           </a>
         <% end %>
-        <h1 class="<%= @article.title_length_classification %>" itemprop="name headline">
-          <%= @article.title %>
```
article title disappeared

juliannatetreault replied:
I've made it reappear! 🧙

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,36 @@
+  {
+    "@context": "http://schema.org",
+    "@type": "Article",
+    "mainEntityOfPage": "True",
```
according to https://schema.org/mainEntityOfPage this should be a property, correct?

rhymes commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,23 +7,34 @@
+  {
+    "@context": "http://schema.org",
+    "@type": "Organization",
+    "mainEntityOfPage": "True",
```
according to https://schema.org/mainEntityOfPage this should be a property, correct?

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,44 @@
+  {
+    "@context": "http://schema.org",
+    "@type": "Person",
+    "mainEntityOfPage": "True",
```
according to https://schema.org/mainEntityOfPage this should be a property, correct?

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,44 @@
+    "image": <%= ProfileImage.new(@user).get(width: 320) %>,
+    "name": "<%= @user.name %>",
+    "description": "<%= @user.summary.presence || ["404 bio not found"].sample %>",
+    "email": "<%= @user.email %>",
```
+1 on hiding this as Ben mentioned

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,44 @@
+    "sameAs": [
+      "https://twitter.com/<%= @user.twitter_username %>",
+      "https://github.com/<%= @user.github_username %>",
+      "<%= @user.mastodon_url %>",
+      "<%= @user.facebook_url %>",
+      "<%= @user.linkedin_url %>",
+      "<%= @user.behance_url %>",
+      "<%= @user.stackoverflow_url %>",
+      "<%= @user.dribbble_url %>",
+      "<%= @user.medium_url %>",
+      "<%= @user.gitlab_url %>",
+      "<%= @user.instagram_url %>",
+      "<%= @user.twitch_url %>",
+      "<%= @user.website_url %>"
```
these could all potentially be empty strings, probably checking ifs as Ben mentioned

juliannatetreault commented:
> This is really cool @juliannatetreault!
> 
> Left some notes :)

Thanks so much @rhymes ! 

I really appreciate all of the suggestions and comments. I believe that I have addressed everything you brought up, but let me know if something still doesn't look quite right. 😊

juliannatetreault commented:
> This is really coming together nicely. Looks like there's a conflict to resolve.
> 
> Good job making turnarounds on the feedback. I'll give it another pass tomorrow.
> 
> In the meantime I think some basic `request` specs which test for simple `expect(response.body).to include "SOME BASIC JSON+LD EXPECTED STRING"` would be good for this PR to ensure it basically acts as expected.
> 
> (E.g. in the `stories_show_spec` etc.)

Thanks a lot @benhalpern !

I believe that I have gone through and addressed all of your comments and made all suggested changes, but let me know if I missed anything. While I haven't pushed up the specs for this yet, they are in the works! 😊

rhymes commented on app/views/articles/_video_player.html.erb:
```diff
@@ -1,16 +1,30 @@
-    <meta itemprop="thumbnailUrl" content="<%= cloudinary(article.video_thumbnail_url, 880) %>" />
-    <meta itemprop="contentUrl" content="<%= article.video_source_url %>" />
+    <meta content="<%= article.published_at %>" />
+    <meta content="<%= article.title %>" />
+    <meta content="<%= article.description %>" />
+    <meta content="<%= cloudinary(article.video_thumbnail_url, 880) %>" />
+    <meta content="<%= article.video_source_url %>" />
     <% minutes, seconds = article.video_duration_in_minutes.split(":") %>
-    <meta itemprop="duration" content="<%= format("PT%<minutes>sM%<seconds>sS", minutes: minutes, seconds: seconds) %>" />
+    <meta content="<%= format("PT%<minutes>sM%<seconds>sS", minutes: minutes, seconds: seconds) %>" />
```
I have the feeling that these meta tags, now that they don't have a `itemprop` anymore, can actually be removed. They lack purpose I guess.

cc @benhalpern

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,53 @@
+      "@context": "http://schema.org",
+      "@type": "Person",
+      "url": "<%= user_url(@user) %>",
+      "name": "<%= @user.name %>",
+      <% if @organization %>
```
unfortunately this is tricky, as if there's no organization this is invalid JSON, see the trailing comma :(

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,53 @@
+    "publisher": {
+      "@context": "http://schema.org",
+      "@type": "Organization",
+      "name": "<%= community_qualified_name %>"
```
missing trailing comma

juliannatetreault replied:
I don't believe that I need a comma here since it is the final key-value pair for `publisher`, but I will double-check to make sure of that. 😊

rhymes replied:
ahahha sorry, I think this was meant for the other file and then I forgot to remove it :D

juliannatetreault replied:
No worries! I appreciate any and all of your feedback. 😄

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,53 @@
     </script>
   <% end %>
 
+<script type="application/ld+json">
```
I tested this after fixing the commas with https://search.google.com/structured-data/testing-tool/u/0/ and it flagged a warning which seems reasonable to me: `The dateModified field is recommended. Please provide a value if available.`. 

Example:

```json
{
    "@context": "http://schema.org",
    "@type": "Article",
    "mainEntityOfPage": {
      "@type": "WebPage",
      "@id": "https://dev.to/http://localhost:3000/funk_damon/nine-coaches-waiting-qui-explicabo-2050"
    },
    "url": "http://localhost:3000/funk_damon/nine-coaches-waiting-qui-explicabo-2050",
    "image": "https://res.cloudinary.com/Optional/image/fetch/s--tRcg02ro--/c_imagga_scale,f_auto,fl_progressive,h_500,q_auto,w_1000/https://pigment.github.io/fake-logos/logos/medium/color/2.png",
    "publisher": {
      "@context": "http://schema.org",
      "@type": "Organization",
      "name": "DEV(local) Community",
      "logo": {
        "@context": "http://schema.org",
        "@type": "ImageObject",
        "url": "https://res.cloudinary.com/Optional/image/fetch/c_scale,fl_progressive,q_auto,w_192/f_auto/https://practicaldev-herokuapp-com.freetls.fastly.net/assets/devlogo-pwa-512.png",
        "width": "192",
        "height": "192"
      }
    },
    "headline": " Nine Coaches Waiting Qui explicabo",
    "author": {
      "@context": "http://schema.org",
      "@type": "Person",
      "url": "http://localhost:3000/funk_damon",
      "name": "Damon Funk",
    },
    "datePublished": "2020-04-01T12:16:23Z"
  }
```

juliannatetreault replied:
I'll make this change! I think I had `dateModified` here previously and removed it. 🤔

rhymes commented on app/views/articles/_video_player.html.erb:
```diff
@@ -1,16 +1,30 @@
   <script src="//content.jwplatform.com/libraries/b1zWy2iv.js" async></script>
   <div id="video-player-<%= article.id %>" class="video-player"></div>
+
+  <script type="application/ld+json">
```
I tested the result of this with https://search.google.com/structured-data/testing-tool/u/0/  and it flagged `A value for the uploadDate field is required.`

Test JSON:

```json
{
  "@context": "https://schema.org",
  "@type": "VideoObject",
  "url": "http://example.com",
  "datePublished": "2020-04-01T12:16:23Z",
  "name": " Nine Coaches Waiting Qui explicabo",
  "description": "Umami selvage hammock. Master fashion axe heirloom.           A   Heirloom try-hard shabby chic. Salv...",
  "thumbnailUrl": "https://dummyimage.com/880",
  "contentUrl": "http://example.com",
  "duration": "PT100M1S"
}
```

so we might want to add that as well.

juliannatetreault replied:
OH! I replaced the `uploadDate` field with `datePublished` on Friday after digging through some docs, but will change this back. 😄

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -85,16 +132,16 @@
-          <meta itemprop="width" content="192">
-          <meta itemprop="height" content="192">
+    <article>
+      <meta content="<%= article_url(@article) %>">
+      <meta content="<%= article_social_image_url(@article) %>">
+      <div>
+        <div>
+          <meta content="<%= cloudinary(SiteConfig.logo_png, 192, "png") %>">
+          <meta content="192">
+          <meta content="192">
         </div>
-        <meta itemprop="name" content="<%= community_qualified_name %>">
+        <meta content="<%= community_qualified_name %>">
```
same as for the video article. I don't think we need these meta fields anymore as they don't have a property

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -120,15 +167,15 @@
-          <span itemprop="author" itemscope itemtype="http://schema.org/Person">
-            <meta itemprop="url" content="<%= user_url(@user) %>">
+          <span>
+            <meta content="<%= user_url(@user) %>">
```
same observation for generic meta tags

rhymes commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,23 +7,38 @@
+      "@id": "https://dev.to/<%= @user.username %>"
+    },
+    "url": "<%= user_url(@user) %>",
+    "image": <%= ProfileImage.new(@user).get(width: 320) %>
```
this is missing the double quotes and the final comma

juliannatetreault replied:
GOOD catch. 😅

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,106 @@
+    "name": "<%= @user.name %>",
+    "description": [
+      {
+        "<%= @user.summary.presence || ["404 bio not found"].sample %>",
```
this is missing the name of the key

juliannatetreault replied:
The name of the key is `description` here. 😊

rhymes replied:
the `description` on line 39 is fine, bu this is technically a hash, see line 40, so it needs a key, but I think I've understood, this is supposed to be an array of strings right? so no, `{ }` is needed. I guess :D

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,106 @@
+      {
+        "<%= @user.summary.presence || ["404 bio not found"].sample %>",
+        "sameAs": [
+          <% if @user.mostly_work_with.present? %>
+            "<%= @user.mostly_work_with %>",
+          <% end %>
+          <% if @user.currently_hacking_on.present? %>
+            "<%= @user.currently_hacking_on %>",
+          <% end %>
+          <% if @user.currently_learning.present? %>
+            "<%= @user.currently_learning %>"
+          <% end %>
```
this is again tricky as if only `mostly_work_with` for example is present, you end up with a an array lke `["value",]` (see the final comma) which is indeed invalid JSON as the testing tool flags.

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,106 @@
-    <meta itemprop="url" content="https://dev.to/<%= @user.username %>">
+
+<script type="application/ld+json">
+  {
```
I wonder if this wouldn't be better built in Ruby in the controller and thenput here as a variable, like

```html
<script type="application/ld-json">
<%= @user_json_ld %>
<script>
```

if you see my comments you'll notice how complicated is to insert if conditon while you're basically building a JSON string. 

I'd build a hash in Ruby and then call `.to_json` in it. 

What do you think?

juliannatetreault replied:
I understand what your concern is and I think that that's not a bad idea!

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,106 @@
+      "@id": "https://dev.to/<%= @user.username %>"
+    },
+    "url": "<%= user_url(@user) %>",
+    "image": <%= ProfileImage.new(@user).get(width: 320) %>,
```
this is missing the double quotes

juliannatetreault commented:
**Rundown of recent changes**:
- added test coverage around templates (users/show, articles/show, _video_player, and _profile_header) to test that the proper data is being rendered 
- cleaned up some of the data structures by adjusting key-value pairs
- extracted the JSON-LD for users/show into a method (`set_json_ld`) in the `stories_controller`, assigned the data to a variable named `user_json_ld`, and placed the method in a `before_action` on the `show` method only (this work is currently commented out)

**Outstanding concerns**:
- RE extraction of JSON-LD into `stories_controller`: I have attempted to handle the JSON in the `stories_controller` to better handle arrays and empty values (per @rhymes 's suggestion), but I am having a difficult time getting the structure to format correctly in the controller. In its current state, the object has a shrinking number of syntax errors that I am attempting to resolve so that I can call the `user_json_ld` variable within the `users/show.html.erb` template in place of the massive JSON-LD object that currently exists. I'm not sure that the way I'm handling the extraction is the best way to do it, so as always, I am very open to feedback on how to improve this. Also, while my gut instinct was to extract this work into the `users_controller`, it seems like the `stories_controller` is what actually handles the `users/show.html.erb` template, which is why I've placed the method there--I'm unsure of whether this is the correct move or not.

- RE removal of `<meta>` tags: I have been going through the process of removing useless `<meta>` tags that once held `itemprop`s, but I have, a couple of times, run into an issue where when I remove a `<meta>` tag,  some of the data does not structure properly in the browser. I suspect this is because I am missing a closing `<div>` tag or the like and I am still working on resolving this.

- RE arrays and JSON-LD: I am running into an issue where some of the data that I've added to the various data structures returns an empty value, even if one is given, e.g. `mostly_works_with`, which I believe is a concern that @rhymes had. After speaking with @benhalpern today, it seems like removing these tricky values and adding them at a later time might be the best approach for this work right now.

All-in-all, I think that everything is coming along nicely and that the outstanding issues can either be handled now or handled in the future when I/we circle back around to JSON-LD work. I appreciate both of your time and patience while working through the wonkiness that comes with implementing JOSN-LD, @rhymes and @benhalpern ! 😄

vaidehijoshi commented on app/controllers/stories_controller.rb:
```diff
@@ -15,6 +15,7 @@ class StoriesController < ApplicationController
 
   before_action :authenticate_user!, except: %i[index search show]
   before_action :set_cache_control_headers, only: %i[index search show]
+  # before_action :set_json_ld, only: :show
```
Something I forgot to mention while we were pairing on this, @juliannatetreault: you might not even need to do this in a before action. Could just set the instance variable directly in the show action:

```
def show
  @json_ld = build_json_ld
end

private
def build_json_ld
  # all your fancy code to build that big object and call `to_json` on it goes here!
end

juliannatetreault replied:
That's not a bad idea; thanks! 😄

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -15,6 +15,7 @@ class StoriesController < ApplicationController
 
   before_action :authenticate_user!, except: %i[index search show]
   before_action :set_cache_control_headers, only: %i[index search show]
+  # before_action :set_json_ld, only: :show
```
Unsure of how everyone feels about extracting the `users/show.html.erb` JSON-LD into the `stories_controller.rb` still - thoughts?

vaidehijoshi replied:
It seems to me like it should still live in `users/show because, from what I could tell, the stories controller redirects to `users/show`.

juliannatetreault replied:
That's my thought as well @vaidehijoshi !

rhymes replied:
I think it's too hard to build a correct JSON with:

```erb
      <% if @user.twitter_username? %>
        "https://twitter.com/<%= @user.twitter_username %>",
      <% end %>
```

as JSON doesn't tolerate trailing commas so you'll always end up with an invalid document if a key is missing or not. 

This is how I would do it:

```ruby
user_json_ld = {
 "mandatory_key_1": "mandatory_value_1",
 "mandatory_key_2": "mandatory_value_2",
  "sameAs": []
}

user_json_ld["sameAs"].append(@user.mastodon_url) if @user.mastodon_url.present?

# and so on...

# once you've built the final Ruby hash we switch to the view
```

```erb
<script type="application/ld+json">
<%= @user_json_ld.to_json.html_safe %>
</script>
```

You'll see that the document this way passes validation :-)

I tried with:

```ruby
    @user_json_ld = {
      "@context": "http://schema.org",
      "@type": "Person",
      "description": "asdasdsa",
      "disambiguatingDescription": [],
      "image": "/uploads/user/profile_image/11/b8bce798-1962-4477-a203-1ea6c35e72b5.png",
      "mainEntityOfPage": {
        "@id": "https://dev.to/rhymes",
        "@type": "WebPage"
      },
      "name": "rhymes",
      "sameAs": [
        "https://github.com/rhymes",
      ],
      "url": "http://localhost:3000/rhymes",
      "worksFor": [
        {
          "@type": "Organization"
        },
      ]
    }
```

ps. don't build it in a before action, as `show` redirects to many different types of pages, there's a `handle_user_index`, you can build it there before `render template: "users/show"`

vaidehijoshi replied:
Agree here too;`handle_user_index` does seem like a better place.

vaidehijoshi commented on app/views/articles/show.html.erb:
```diff
@@ -85,16 +133,14 @@
+          <% cloudinary(SiteConfig.logo_png, 192, "png") %>
         </div>
-        <meta itemprop="name" content="<%= community_qualified_name %>">
+        <% community_qualified_name %>
```
```suggestion
        <%= community_qualified_name %>
```
should this be invoked here? looking at the diff, it seems like we used to do this: `<%= community_qualified_name %>">`, so i'm confused why we're not doing `<%= community_qualified_name %>`

juliannatetreault replied:
Good question! I actually meant to add this to my list of concerns. When I remove the `<meta>` tag that was wrapped around the `erb`, the code within the `erb` renders on the webpage as text, as shown in the before and after pictures below. 

Two possible solutions that I've found so far are to either 1.) not invoke the `erb` or 2.) wrap the `erb` in `<a>` tags in place of the `<meta>` tags. I don't think that either of the two solutions is great and would like more opinions on this or possible solutions for it. 😄 

With `<meta>` tags removed and without invoking the `erb`:
![Screen Shot 2020-04-09 at 1 15 44 PM](https://user-images.githubusercontent.com/32834804/78932627-1d4b9680-7a65-11ea-9599-7677625d5d13.png)


With the `<meta>` tags removed, while still invoking the `erb`:
![Screen Shot 2020-04-09 at 1 16 16 PM](https://user-images.githubusercontent.com/32834804/78932815-6bf93080-7a65-11ea-9db1-f35d453ea976.png)

vaidehijoshi replied:
Ah, when I first looked at this diff, I didn't actually catch that this was inside of the `<meta>`. My next question would be: is removing the `<meta>` tag intentional??

If it _is_ and we _do_ want to remove it, then you should just remove the `<% community_qualified_name %>`; you aren't invoking it, so it's on the template but isn't being used at all.

`<% community_qualified_name %>` is just executing the variable but isn't rendering it, so it's wasted effort because that variable is thrown away.

vaidehijoshi replied:
> When I remove the <meta> tag that was wrapped around the erb, the code within the erb renders on the webpage as text, as shown in the before and after pictures below.

This is expected, because you're just actually just invoking what's inside the erb, you're actually rendering it! :) In one situation you have an expression, in another situation you have a scriptlet, as described [here](https://www.stuartellis.name/articles/erb).

juliannatetreault replied:
OH, I see; that makes sense! To answer your question, yes, removing the `<meta>` tag was intentional, but I was hesitant to remove the `erb` from the template after removing the `<meta>` tags because I wasn't sure if I was removing too much, but now that you mention it, it seems like it's actually just dead code if I leave the `erb` sans `<meta>` tags in there.

vaidehijoshi commented on app/views/articles/show.html.erb:
```diff
@@ -120,15 +166,15 @@
-          <span itemprop="author" itemscope itemtype="http://schema.org/Person">
-            <meta itemprop="url" content="<%= user_url(@user) %>">
+          <span>
+            <% user_url(@user) %>
```
same question here; why not `<%= user_url(@user) %>`. please excuse my ignorance if i'm missing something totally obvious here, though...

juliannatetreault replied:
No worries! See above. 😄

vaidehijoshi replied:
Same question/suggestion as above here. As the erb currently is written, this is being invoked in the template but then the output is just being thrown away (not rendered).

vaidehijoshi commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+      "alumniOf": "<%= @user.education %>"
+    <% end %>
+  }
+</script>
```
how did it go with continuing to pull this out into an instance variable and rendering it via the controller? did you run into issues with that?

juliannatetreault replied:
I still ran into a litany of syntax errors, making it not usable for the time being. But, as I mentioned in the comment in the `stories_controller.rb` and in the comment on the PR, I'm still open to trying to get it to work if the consensus is that it is better to extract it into the controller. 😄

vaidehijoshi replied:
yeah, i saw the comments, but i was specifically asking because i wasn't sure if you were close to making it work or if it ended up being way more trouble than it was worth. i saw the commented out method but wasn't sure whether you left it in for reference or because you planned to use it later after getting feedback

juliannatetreault replied:
Ah, gotcha! It was definitely a fair amount of trouble, but I left the comment there as both a reference and a shortcut for myself if we decide to go the extraction direction.

vaidehijoshi commented on spec/requests/articles/_video_player_show_spec.rb:
```diff
@@ -0,0 +1,43 @@
+    it "renders the proper modified at date" do
+      time_now = Time.current
+      video_article.edited_at = time_now
+      expect(response.body).to include CGI.escapeHTML(video_article.readable_edit_date)
```
i'm curious: how do we know that `readable_edit_date` is the correct date? should we be checking that `video_article .readable_edit_date` is the same as `time_now`?

vaidehijoshi replied:
from looking at this test, it's not clear to me why we even set the video's `edited_at` to a new time if we don't ever check it in the `expect` block...

juliannatetreault replied:
Hm, I see how that could be unclear. I have since changed this for both the published and modified dates in both this spec and the `articles_show_spec.rb`. In addition to that, I've removed the modified date test from this spec because I don't actually have the modified date in the JSON-LD videoObject. 😅

vaidehijoshi commented on spec/requests/articles/articles_show_spec.rb:
```diff
@@ -0,0 +1,41 @@
+    it "renders the proper modified at date" do
+      time_now = Time.current
+      article.edited_at = time_now
+      expect(response.body).to include CGI.escapeHTML(article.readable_edit_date)
```
same question about this one, if i'm missing some context here, again, please do let me know!

juliannatetreault replied:
I've since changed this to make it more reviewer/future programmer-friendly. Let me know what you think. 😄

vaidehijoshi commented on spec/requests/user/user_show_spec.rb:
```diff
@@ -0,0 +1,52 @@
+require "rails_helper"
+
+RSpec.describe "UserShow", type: :request do
+  let_it_be(:user) { create(:user, email_public: true, employment_title: "SEO", employer_name: "DEV", currently_hacking_on: "JSON-LD", education: "DEV University", linkedin_url: "www.linkedin.com/company/example/") }
```
might want to split this up into separate lines to make it easier to read/parse

vaidehijoshi commented on spec/requests/user/user_show_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+    it "renders the proper additional username for a user when one is present" do
+      user.update(github_username: "username")
+      expect(response.body).to include CGI.escapeHTML(user.github_username)
```
same question here: should we testing that the updated value is what appears in the escaped HTML?

```suggestion
      expect(response.body).to include CGI.escapeHTML("username")
```

juliannatetreault replied:
Good question! I shouldn't be updating the GitHub username here since it already exists--I should just be checking that the user's GitHub username is in the response body.

juliannatetreault commented on app/views/articles/_video_player.html.erb:
```diff
@@ -1,16 +1,28 @@
-    <div id="video-player-source" data-source="<%= article.video_source_url %>"></div>
-  <% end %>
+<div class="video-player-header">
+  <%= article.published_at %>
+  <%= article.title %>
+  <%= article.description %>
+  <%= cloudinary(article.video_thumbnail_url, 880) %>
+  <%= article.video_source_url %>
```
I suppose I should be removing this as well as line 8 since they were all `erb` in meta tags as well, correct @vaidehijoshi ?

vaidehijoshi replied:
Confirm. It looks like they were previously being rendered dynamically within the `<meta>` tags, but since those are now removed, you're now just rendering them on the page. In fact, if you look at the vide player page locally, you'll see all this stuff rendered on the page and I'd guess that it breaks the way the view is rendered 😬So, you are right to remove them unless you intentionally wanted to render those on the view player page

vaidehijoshi commented on app/views/articles/show.html.erb:
```diff
@@ -199,6 +236,7 @@
     <%= render "articles/org_branding", organization: @organization %>
     <%= render "articles/full_comment_area" %>
   </div>
+
```
Intentional empty line?

vaidehijoshi commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+    ],
+    "image": "<%= ProfileImage.new(@user).get(width: 320) %>",
+    "name": "<%= @user.name %>",
+    <% if @user.email_public == true %>
```
You can just do `@user.email_public` here because it returns a Boolean value. The `== true` is redundant

juliannatetreault replied:
I had added this as a test and never removed this. 🤦‍♀️ Good catch!

vaidehijoshi commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+    <% if @user.employment_title.present? %>
+      "jobTitle": "<%= @user.employment_title %>",
+    <% end %>
+    "description": "<%= @user.summary.presence || ["404 bio not found"].sample %>",
```
Curious why `presence` here when elsewhere you’re using `present?`

juliannatetreault replied:
`presence` was already used for the description check and then I added the `.present?` checks to the other conditionals much later--I will make this more consistent!

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,54 @@
+      "url": "<%= user_url(@user) %>",
+      "name": "<%= @user.name %>",
+      <% if @organization.present? %>
+        "organization": {
```
The validator says the following:

![Screenshot_2020-04-11 Structured Data Testing Tool](https://user-images.githubusercontent.com/146201/79044806-997ddf80-7c07-11ea-821e-3492dd83cf73.png)

rhymes commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,23 +7,38 @@
+    "@type": "Organization",
+    "mainEntityOfPage": {
+      "@type": "WebPage",
+      "@id": "https://dev.to/<%= @user.username %>"
```
we should use `user_url(@user)` here

rhymes commented on app/views/users/_profile_header.html.erb:
```diff
@@ -7,23 +7,38 @@
+</script>
+
+<div class="user-profile-header" style="<%= user_colors_style(@user) %>">
+  <a href="https://dev.to/<%= @user.username %>"></a>
```
we should use `user_url(@user)` here

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+    "@type": "Person",
+    "mainEntityOfPage": {
+      "@type": "WebPage",
+      "@id": "https://dev.to/<%= @user.username %>"
```
we should use `user_url(@user)` here

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+      <% if @user.instagram_url? %>
+        "<%= @user.instagram_url %>",
+      <% end %>
+      <% if @user.twitch_username? %>
+        "<%= @user.twitch_url %>",
+      <% end %>
+      <% if @user.website_url? %>
+        "<%= @user.website_url %>"
+      <% end %>
```
this won't work, because if twitch_username is present, but not the website (or any of the combinations before) the array will have a trailing comma, which is invalid in JSON (you're building a JSON document, not a Ruby hash).

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+    <% end %>
+    "description": "<%= @user.summary.presence || ["404 bio not found"].sample %>",
+    "disambiguatingDescription": [
+      <% if @user.mostly_work_with.presence %>
```
`.presence` returns the value, you're probably better off with `.present?` which returns true or false, not a big deal though

vaidehijoshi replied:
I agree with @rhymes on this one

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+    "worksFor": [
+      {
+        "@type": "Organization",
+        <% if @user.employer_name.presence %>
+          "name": "<%= @user.employer_name %>",
+        <% end %>
+        <% if @user.employer_url.presence %>
+          "sameAs": [ "<%= @user.employer_url %>" ]
+        <% end %>
```
same problem as most of the other if, if it end ups with both of these not being present, the document will be invalid because of the comm after `Organization`. Same if `employer_url` missing but `employer_name` being present

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+        "<%= @user.currently_learning %>"
+      <% end %>
+      ],
+    "worksFor": [
```
maybe this shouldn't be present at all if both `employer_name` and `employer_url` are missing?

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+        <% end %>
+      },
+    ],
+    <% if @user.education.presence %>
+      "alumniOf": "<%= @user.education %>"
+    <% end %>
```
this not being present will lead to a trailing command on line 108 and the document being invalid

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+        <% if @user.employer_url.presence %>
+          "sameAs": [ "<%= @user.employer_url %>" ]
+        <% end %>
+      },
```
no trailing comma here

rhymes commented on app/views/users/show.html.erb:
```diff
@@ -24,15 +24,103 @@
+</script>
+
+  <div class="user-profile-header" style="<%= user_colors_style(@user) %>">
+    <a href="https://dev.to/<%= @user.username %>"></a>
```
this should be `user_url(@user)`

rhymes commented on spec/requests/articles/_video_player_show_spec.rb:
```diff
@@ -0,0 +1,37 @@
+require "rails_helper"
```
the file name shouldn't have a leading undescore?

rhymes commented on spec/requests/user/_profile_header_spec.rb:
```diff
@@ -0,0 +1,27 @@
+require "rails_helper"
```
I don't think this file is needed as there's already a `user_show_spec.rb`. After all you're testing the page, not the view partial

juliannatetreault commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,54 @@
+      "@context": "http://schema.org",
+      "@type": "Person",
+      "url": "<%= user_url(@user) %>",
+      "name": "<%= @user.name %>",
```
Hmm, I _think_ I see what you mean! I've made a change to this, but I'm not sure that it is any better than what it was previously... 🤔

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,54 @@ def assign_podcasts
+    user_same_as
+  end
+
+  def user_same_as
+    @user_json_ld["sameAs"]&.append(@user.mastodon_url) if @user.mastodon_url.present?
+    @user_json_ld["sameAs"]&.append(@user.facebook_url) if @user.facebook_url.present?
+    @user_json_ld["sameAs"]&.append(@user.linkedin_url) if @user.linkedin_url.present?
+    @user_json_ld["sameAs"]&.append(@user.behance_url) if @user.behance_url.present?
+    @user_json_ld["sameAs"]&.append(@user.stackoverflow_url) if @user.stackoverflow_url.present?
+    @user_json_ld["sameAs"]&.append(@user.dribbble_url) if @user.dribbble_url.present?
+    @user_json_ld["sameAs"]&.append(@user.medium_url) if @user.medium_url.present?
+    @user_json_ld["sameAs"]&.append(@user.gitlab_url) if @user.gitlab_url.present?
+    @user_json_ld["sameAs"]&.append(@user.instagram_url) if @user.instagram_url.present?
+    @user_json_ld["sameAs"]&.append(@user.twitch_username) if @user.twitch_username.present?
+    @user_json_ld["sameAs"]&.append(@user.website_url) if @user.website_url.present?
+  end
```
This method exists because I ran into a `Metrics/CyclomaticComplexity(RuboCop)` error when I initially tried adding this code into the `set_json_ld` method. This is also why the `twitter_username` and `github_username` exist in the `set_json_ld` and not in the `user_same_as` method. 😄

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld["sameAs"]&.append(@user.github_username) if @user.github_username.present?
+  end
+
+  def user_same_as_json_ld
+    @user_json_ld["sameAs"]&.append(@user.mastodon_url) if @user.mastodon_url.present?
+    @user_json_ld["sameAs"]&.append(@user.facebook_url) if @user.facebook_url.present?
+    @user_json_ld["sameAs"]&.append(@user.linkedin_url) if @user.linkedin_url.present?
+    @user_json_ld["sameAs"]&.append(@user.behance_url) if @user.behance_url.present?
+    @user_json_ld["sameAs"]&.append(@user.stackoverflow_url) if @user.stackoverflow_url.present?
+    @user_json_ld["sameAs"]&.append(@user.dribbble_url) if @user.dribbble_url.present?
+    @user_json_ld["sameAs"]&.append(@user.medium_url) if @user.medium_url.present?
+    @user_json_ld["sameAs"]&.append(@user.gitlab_url) if @user.gitlab_url.present?
+    @user_json_ld["sameAs"]&.append(@user.instagram_url) if @user.instagram_url.present?
+    @user_json_ld["sameAs"]&.append(@user.twitch_username) if @user.twitch_username.present?
+    @user_json_ld["sameAs"]&.append(@user.website_url) if @user.website_url.present?
+  end
```
This method exists because I ran into a Metrics/CyclomaticComplexity(RuboCop) error when I initially tried adding this code into the `set_json_ld` method. This is also why the `twitter_username` and `github_username` exist in the `user_profile_details_json_ld` and not in the `user_same_as_json_ld` method. 😄

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    user_same_as_json_ld
+  end
+
+  def user_profile_details_json_ld
+    @user_json_ld["disambiguatingDescription"]&.append(@user.mostly_work_with) if @user.mostly_work_with.present?
+    @user_json_ld["disambiguatingDescription"]&.append(@user.currently_hacking_on) if @user.currently_hacking_on.present?
+    @user_json_ld["disambiguatingDescription"]&.append(@user.currently_learning) if @user.currently_learning.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_name) if @user.employer_name.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_url) if @user.employer_url.present?
+    @user_json_ld["alumniOf"]&.append(@user.education) if @user.education.present?
+    @user_json_ld["email"]&.append(@user.email) if @user.email_public
+    @user_json_ld["jobTitle"]&.append(@user.employment_title) if @user.employment_title.present?
+    @user_json_ld["sameAs"]&.append(@user.twitter_username) if @user.twitter_username.present?
+    @user_json_ld["sameAs"]&.append(@user.github_username) if @user.github_username.present?
+  end
```
This method exists to resolve Code Climate issues--if I don't extract this logic out of the `set_json_ld` method, then the method is over 25 lines long, causing a Code Climate failure.

vaidehijoshi commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+      ],
+      "alumniOf": ""
+    }
+    user_profile_details_json_ld
+    user_same_as_json_ld
```
Since these methods are actually modifying the instance of `@user_json_ld`, you might want to consider renaming the method names to suggest that. Maybe `set_user_profile_json_ld` and `set_user_same_as_json_ld`? Or something less verbose if can come up with it ;)

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    set_user_same_as_json_ld
+  end
+
+  def set_user_profile_json_ld
+    @user_json_ld["disambiguatingDescription"]&.append(@user.mostly_work_with) if @user.mostly_work_with.present?
+    @user_json_ld["disambiguatingDescription"]&.append(@user.currently_hacking_on) if @user.currently_hacking_on.present?
+    @user_json_ld["disambiguatingDescription"]&.append(@user.currently_learning) if @user.currently_learning.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_name) if @user.employer_name.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_url) if @user.employer_url.present?
+    @user_json_ld["alumniOf"]&.append(@user.education) if @user.education.present?
+    @user_json_ld["email"]&.append(@user.email) if @user.email_public
+    @user_json_ld["jobTitle"]&.append(@user.employment_title) if @user.employment_title.present?
+    @user_json_ld["sameAs"]&.append(@user.twitter_username) if @user.twitter_username.present?
+    @user_json_ld["sameAs"]&.append(@user.github_username) if @user.github_username.present?
+  end
```
This method exists to resolve Code Climate issues--if I don't extract this logic out of the `set_json_ld` method, then the method is over 25 lines long, causing a Code Climate failure.

juliannatetreault commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld["sameAs"]&.append(@user.github_username) if @user.github_username.present?
+  end
+
+  def set_user_same_as_json_ld
+    @user_json_ld["sameAs"]&.append(@user.mastodon_url) if @user.mastodon_url.present?
+    @user_json_ld["sameAs"]&.append(@user.facebook_url) if @user.facebook_url.present?
+    @user_json_ld["sameAs"]&.append(@user.linkedin_url) if @user.linkedin_url.present?
+    @user_json_ld["sameAs"]&.append(@user.behance_url) if @user.behance_url.present?
+    @user_json_ld["sameAs"]&.append(@user.stackoverflow_url) if @user.stackoverflow_url.present?
+    @user_json_ld["sameAs"]&.append(@user.dribbble_url) if @user.dribbble_url.present?
+    @user_json_ld["sameAs"]&.append(@user.medium_url) if @user.medium_url.present?
+    @user_json_ld["sameAs"]&.append(@user.gitlab_url) if @user.gitlab_url.present?
+    @user_json_ld["sameAs"]&.append(@user.instagram_url) if @user.instagram_url.present?
+    @user_json_ld["sameAs"]&.append(@user.twitch_username) if @user.twitch_username.present?
+    @user_json_ld["sameAs"]&.append(@user.website_url) if @user.website_url.present?
+  end
```
This method exists because I ran into a `Metrics/CyclomaticComplexity(RuboCop)` error when I initially tried adding this code into the `set_json_ld` method. This is also why the `twitter_username` and `github_username` exist in the `set_user_profile_json_ld` and not in the `set_user_same_as_json_ld` method. 😄

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+  end
+
+  def set_user_profile_json_ld
+    @user_json_ld["disambiguatingDescription"]&.append(@user.mostly_work_with) if @user.mostly_work_with.present?
```
you can remove the `&` as you already know the key exists and it's an array. Same for the other similar keys

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld["disambiguatingDescription"]&.append(@user.currently_learning) if @user.currently_learning.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_name) if @user.employer_name.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_url) if @user.employer_url.present?
+    @user_json_ld["alumniOf"]&.append(@user.education) if @user.education.present?
```
`alumniOf` on line 356 is a strig, not an array

juliannatetreault replied:
🤦‍♀️

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld["worksFor"]&.append(@user.employer_name) if @user.employer_name.present?
+    @user_json_ld["worksFor"]&.append(@user.employer_url) if @user.employer_url.present?
+    @user_json_ld["alumniOf"]&.append(@user.education) if @user.education.present?
+    @user_json_ld["email"]&.append(@user.email) if @user.email_public
```
`email` is a string, not an array, see line 347

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld["worksFor"]&.append(@user.employer_url) if @user.employer_url.present?
+    @user_json_ld["alumniOf"]&.append(@user.education) if @user.education.present?
+    @user_json_ld["email"]&.append(@user.email) if @user.email_public
+    @user_json_ld["jobTitle"]&.append(@user.employment_title) if @user.employment_title.present?
```
`jobTitle` is a string, not an array, see line 348

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+  end
+
+  def set_user_same_as_json_ld
+    @user_json_ld["sameAs"]&.append(@user.mastodon_url) if @user.mastodon_url.present?
```
you can remove eth `&` as you alread know the key is present and an array

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,40 @@
+    "@type": "Article",
+    "mainEntityOfPage": {
+      "@type": "WebPage",
+      "@id": "https://dev.to/<%= article_url(@article) %>"
```
article URL alreaady returns the URL, it wuld be invalid with the prefix :)

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -31,6 +31,40 @@
+      "name": "<%= @user.name %>"
+    },
+    "datePublished": "<%= @article.published_timestamp %>",
+    "dateModified": "<%= @article.edited_at || @article.published_timestamp %>"
```
If you look at the implementation of `published_timestamp`, we return the ISO date so here we have to do the same, otherwise the date will have a different format, for example:

```ruby
[2] pry(main)> ar.created_at.to_s
=> "2020-04-01 12:16:31 UTC"
[3] pry(main)> ar.created_at.iso8601
=> "2020-04-01T12:16:31Z"
```

I suggest `@article.edited_at&.iso8601 || @article.published_timestamp`

juliannatetreault replied:
Good catch! 👍

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+      "@type": "Person",
+      "mainEntityOfPage": {
+        "@type": "WebPage",
+        "@id": "<%= user_url(@user) %>"
```
Since now we're writing Ruby, this won't be interpreted by the ERB (the template) parser, so it'll just be a string with tha value, see: 

![Screenshot_2020-04-15 Structured Data Testing Tool](https://user-images.githubusercontent.com/146201/79337300-8760b180-7f25-11ea-9708-c057cc931877.png)

this for example has to become: `"@id": URL.user(@user)`

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+        "@type": "WebPage",
+        "@id": "<%= user_url(@user) %>"
+      },
+      "url": "<%= user_url(@user) %>",
```
same `URL.user(@user)`

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+      },
+      "url": "<%= user_url(@user) %>",
+      "sameAs": [],
+      "image": "<%= ProfileImage.new(@user).get(width: 320) %>",
```
`"<%= ProfileImage.new(@user).get(width: 320) %>"` becomes `ProfileImage.new(@user).get(width: 320)`

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+      "url": "<%= user_url(@user) %>",
+      "sameAs": [],
+      "image": "<%= ProfileImage.new(@user).get(width: 320) %>",
+      "name": "<%= @user.name %>",
```
just `@user.name`

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+      "name": "<%= @user.name %>",
+      "email": "",
+      "jobTitle": "",
+      "description": "<%= @user.summary.presence || ['404 bio not found'].sample %>",
```
same logic as the others

rhymes commented on app/controllers/stories_controller.rb:
```diff
@@ -330,4 +331,58 @@ def assign_podcasts
+    @user_json_ld[:disambiguatingDescription].append(@user.currently_learning) if @user.currently_learning.present?
+    @user_json_ld[:worksFor][0][:employer_name] = @user.employer_name if @user.employer_name.present?
+    @user_json_ld[:worksFor][0][:employer_url] = @user.employer_url if @user.employer_url.present?
+    @user_json_ld[:alumniOf].concat(@user.education) if @user.education.present?
+    @user_json_ld[:email].concat(@user.email) if @user.email_public
+    @user_json_ld[:jobTitle].concat(@user.employment_title) if @user.employment_title.present?
```
these are just string keys, you can use, for example: 

```ruby
@user_json_ld[:alumniOf] = @user.education if @user.education.present?
```

no need for concat

rhymes commented on db/schema.rb:
```diff
@@ -307,7 +307,6 @@
     t.string "slug"
     t.string "status", default: "active"
     t.datetime "updated_at", null: false
-    t.index ["slug"], name: "index_chat_channels_on_slug", unique: true
```
this stays in :)

juliannatetreault replied:
That snuck in! 😄
